<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Unsupervised-Learning-Guided-Lesson" data-toc-modified-id="Unsupervised-Learning-Guided-Lesson-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Unsupervised Learning Guided Lesson</a></span><ul class="toc-item"><li><span><a href="#Lesson-Goals" data-toc-modified-id="Lesson-Goals-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Lesson Goals</a></span></li><li><span><a href="#Exploring-the-Variables" data-toc-modified-id="Exploring-the-Variables-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Exploring the Variables</a></span></li><li><span><a href="#Some-More-Transformations---PCA" data-toc-modified-id="Some-More-Transformations---PCA-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Some More Transformations - PCA</a></span></li><li><span><a href="#The-Algorithm---K-Means-Clustering" data-toc-modified-id="The-Algorithm---K-Means-Clustering-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>The Algorithm - K-Means Clustering</a></span></li></ul></li></ul></div>

# Unsupervised Learning Guided Lesson


## Lesson Goals

In this guided lesson, we will analyze an unsupervised learning problem from start to finish and introduce several different processing techniques.


**Introduction**

As a data scientist or analyst, you may be asked open ended question about a dataset. One example is to find some patterns in a dataset that is unlabeled. Typically, this happens when analyzing a group of customers and trying to find a common themes between the transactions. The leading choice of algorithm for this type of problem is an unsupervised algorithm. Specifically, this lesson will be using clustering to analyze this problem. In this lesson, we will be analyzing a log of transactions from a bakery to make recommendations about the marketing and sale of products.


**The Data**

The dataset we will be analyzing comes from Kaggle and is a log of transactions from a bakery in Edinburgh called The Bread Basket. We will start by evaluating the types of the variables in the data as well as the contents of the categorical variables and the distribution of the numerical variables. 

In [21]:
%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

breadbasket = pd.read_csv(r'C:\Users\happy\Documents\ironhack\RecoveryPlan\Week5\lab-unsupervised-learning-en\data\BreadBasket_DMS.csv')
breadbasket.dtypes

Date           object
Time           object
Transaction     int64
Item           object
dtype: object

In [22]:
breadbasket.head()

Date      Time  Transaction           Item
0  2016-10-30  09:58:11            1          Bread
1  2016-10-30  10:05:34            2   Scandinavian
2  2016-10-30  10:05:34            2   Scandinavian
3  2016-10-30  10:07:57            3  Hot chocolate
4  2016-10-30  10:07:57            3            Jam

In [23]:
breadbasket.describe()

Transaction
count  21293.000000
mean    4951.990889
std     2787.758400
min        1.000000
25%     2548.000000
50%     5067.000000
75%     7329.000000
max     9684.000000

As we can see from these three functions, we have 4 columns in the dataset. The date and time are separated into two columns and stored as text. This means we will have to combine and convert them later on. The transaction variable is ordinal, so the summary statistics really have no meaning in this context. The only meaningful information from the describe function is the max which tells us we have 9684 transactions in the dataset. From the head function, we see that each row represents an item in the transaction. This means that there are potentially multiple items per transaction or maybe just one. We might benefit from consolidating the data and creating a new dataset that contains one row per transaction.

So the data we have is a breakdown of each item in a transaction and the date and time when the transaction occurred. The data we do not have is any information about the customer. Since we cannot associate the transactions back to customers, we cannot tell if a certain customer is a regular who comes in and buys a coffee and a pastry every day or whether a customer is a tourist who came in once and bought a specialty dessert.

Our strategy to make sense of the dataset will be to generate derived variables from this transaction log and cluster based on these derived variables. We will evaluate the aggregate information regarding each cluster and make recommendations about which products to advertise and which products should be in stock and at what days and times.


## Exploring the Variables

While we only have a few variables, we should explore their contents.

First, let's look at the time and day of week. Hour is a crucial factor since customer behavior differs significantly between the morning and the afternoon. However, we can even find differences between customer behavior at 7 am vs. at 9 am. Similarly, we see differences between weekday and weekend customer behavior.

To examine the date and time, we must reformat this variable. We start by combining the date and time into one column and then transforming the column to a datetime column. This allows us to extract the hour and the time of day.

In [24]:
breadbasket['DateTime'] = pd.to_datetime(breadbasket.Date + ' ' + breadbasket.Time)

Next, we look at the item variable. This variable will tell us how many products are sold by the bakery and which products are more popular.

In [25]:
breadbasket.Item.unique()

array(['Bread', 'Scandinavian', 'Hot chocolate', 'Jam', 'Cookies',
       'Muffin', 'Coffee', 'Pastry', 'Medialuna', 'Tea', 'NONE',
       'Tartine', 'Basket', 'Mineral water', 'Farm House', 'Fudge',
       'Juice', "Ella's Kitchen Pouches", 'Victorian Sponge', 'Frittata',
       'Hearty & Seasonal', 'Soup', 'Pick and Mix Bowls', 'Smoothies',
       'Cake', 'Mighty Protein', 'Chicken sand', 'Coke',
       'My-5 Fruit Shoot', 'Focaccia', 'Sandwich', 'Alfajores', 'Eggs',
       'Brownie', 'Dulce de Leche', 'Honey', 'The BART', 'Granola',
       'Fairy Doors', 'Empanadas', 'Keeping It Local', 'Art Tray',
       'Bowl Nic Pitt', 'Bread Pudding', 'Adjustment', 'Truffles',
       'Chimichurri Oil', 'Bacon', 'Spread', 'Kids biscuit', 'Siblings',
       'Caramel bites', 'Jammie Dodgers', 'Tiffin', 'Olum & polenta',
       'Polenta', 'The Nomad', 'Hack the stack', 'Bakewell',
       'Lemon and coconut', 'Toast', 'Scone', 'Crepes', 'Vegan mincepie',
       'Bare Popcorn', 'Muesli', 'Crisps', 'Pi

We can look at the counts as well to see what items are most popular.

In [26]:
breadbasket.Item.value_counts()

Item
Coffee            5471
Bread             3325
Tea               1435
Cake              1025
Pastry             856
                  ... 
Chicken sand         1
The BART             1
Adjustment           1
Olum & polenta       1
Bacon                1
Name: count, Length: 95, dtype: int64

There are a total of 95 items. The most common items are coffee, bread, tea, cake and pastry. Since 95 is a really large number, if we created dummy variables out of this data, it would produce too many variables. One option is to classify the data. Typically, when working on these types of problems, companies will have a classification system for the items they sell. However, let's try to come up with our own.

Looking at the list of unique items, we can identify a number of obvious categories. We have beverages and breakfast pastries like muffins and medialuna. We have items for kids like juice and pouches. We also have non food items like gift vouchers and t-shirts. Another group of items that we can notice is ready to eat snacks like popcorn and crisps. With a bit of work, we can narrow it down from 95 products to 11 categories: beverage, other, kids, snacks, bread, breakfast pastry, dessert, condiments, breakfast, lunch, and other foods. The last group is used to classify mostly uncommon items that sell very little (like polenta) or have names that are hard to identify (like "Hack the Stack").

We generate the categories using lists and then use the lists to create dummy variables.



In [27]:
beverage = ['Hot chocolate', 'Coffee', 'Tea', 'Mineral water', 'Juice', 'Coke', 'Smoothies']
other = ['NONE', 'Christmas common', 'Gift voucher', "Valentine's card", 'Tshirt', 'Afternoon with the baker', 'Postcard', 'Siblings', 'Nomad bag', 'Adjustment', 'Drinking chocolate spoons ', 'Coffee granules ']
kids = ["Ella's Kitchen Pouches", 'My-5 Fruit Shoot', 'Kids biscuit']
snacks = ['Mighty Protein', 'Pick and Mix Bowls', 'Caramel bites', 'Bare Popcorn', 'Crisps', 'Cherry me Dried fruit', 'Raw bars']
bread = ['Bread', 'Toast', 'Baguette', 'Focaccia', 'Scandinavian']
breakfast_pastry = ['Muffin', 'Pastry', 'Medialuna', 'Scone']
dessert = ['Cookies', 'Tartine', 'Fudge', 'Victorian Sponge', 'Cake', 'Alfajores', 'Brownie', 'Bread Pudding', 'Bakewell', 'Raspberry shortbread sandwich', 'Lemon and coconut', 'Crepes', 'Chocolates', 'Truffles', 'Panatone']
condiments = ['Jam', 'Dulce de Leche', 'Honey', 'Gingerbread syrup', 'Extra Salami or Feta', 'Bacon', 'Spread', 'Chimichurri Oil']
breakfast = ['Eggs', 'Frittata', 'Granola', 'Muesli', 'Duck egg', 'Brioche and salami']
lunch = ['Soup', 'Sandwich', 'Chicken sand', 'Salad', 'Chicken Stew']


other_food = [x for x in breadbasket.Item.unique() if x not in beverage 
                and x not in other and x not in kids and x not in snacks 
                and x not in bread and x not in breakfast_pastry 
                and x not in dessert and x not in condiments 
                and x not in breakfast and x not in lunch]


breadbasket['beverage'] = np.where(breadbasket.Item.isin(beverage), 1, 0)
breadbasket['other'] = np.where(breadbasket.Item.isin(other), 1, 0)
breadbasket['kids'] = np.where(breadbasket.Item.isin(kids), 1, 0)
breadbasket['snacks'] = np.where(breadbasket.Item.isin(snacks), 1, 0)
breadbasket['bread'] = np.where(breadbasket.Item.isin(bread), 1, 0)
breadbasket['breakfast_pastry'] = np.where(breadbasket.Item.isin(breakfast_pastry), 1, 0)
breadbasket['dessert'] = np.where(breadbasket.Item.isin(dessert), 1, 0)
breadbasket['condiments'] = np.where(breadbasket.Item.isin(condiments), 1, 0)
breadbasket['breakfast'] = np.where(breadbasket.Item.isin(breakfast), 1, 0)
breadbasket['lunch'] = np.where(breadbasket.Item.isin(lunch), 1, 0)
breadbasket['other_food'] = np.where(breadbasket.Item.isin(other_food), 1, 0)
breadbasket.head()

Date      Time  Transaction           Item            DateTime  \
0  2016-10-30  09:58:11            1          Bread 2016-10-30 09:58:11   
1  2016-10-30  10:05:34            2   Scandinavian 2016-10-30 10:05:34   
2  2016-10-30  10:05:34            2   Scandinavian 2016-10-30 10:05:34   
3  2016-10-30  10:07:57            3  Hot chocolate 2016-10-30 10:07:57   
4  2016-10-30  10:07:57            3            Jam 2016-10-30 10:07:57   

   beverage  other  kids  snacks  bread  breakfast_pastry  dessert  \
0         0      0     0       0      1                 0        0   
1         0      0     0       0      1                 0        0   
2         0      0     0       0      1                 0        0   
3         1      0     0       0      0                 0        0   
4         0      0     0       0      0                 0        0   

   condiments  breakfast  lunch  other_food  
0           0          0      0           0  
1           0          0      0           0  
2           0          0      0           0  
3           0          0      0           0  
4           1          0      0           0

**Processing the Data**

The first bit of work we will do to process the data is to aggregate by transaction. This will give us the count of each category per transaction. We will use the groupby function to find the sum in each transaction. We group by the datetime as well since we want to keep this column after the aggregation. This should not be a problem since a transaction number and a datetime uniquely identifies each row. 

In [28]:
bread_group = breadbasket.groupby(['Transaction','DateTime']).sum()
bread_group.head()

Date  \
Transaction DateTime                                              
1           2016-10-30 09:58:11                      2016-10-30   
2           2016-10-30 10:05:34            2016-10-302016-10-30   
3           2016-10-30 10:07:57  2016-10-302016-10-302016-10-30   
4           2016-10-30 10:08:41                      2016-10-30   
5           2016-10-30 10:13:03  2016-10-302016-10-302016-10-30   

                                                     Time  \
Transaction DateTime                                        
1           2016-10-30 09:58:11                  09:58:11   
2           2016-10-30 10:05:34          10:05:3410:05:34   
3           2016-10-30 10:07:57  10:07:5710:07:5710:07:57   
4           2016-10-30 10:08:41                  10:08:41   
5           2016-10-30 10:13:03  10:13:0310:13:0310:13:03   

                                                     Item  beverage  other  \
Transaction DateTime                                                         
1           2016-10-30 09:58:11                     Bread         0      0   
2           2016-10-30 10:05:34  ScandinavianScandinavian         0      0   
3           2016-10-30 10:07:57   Hot chocolateJamCookies         1      0   
4           2016-10-30 10:08:41                    Muffin         0      0   
5           2016-10-30 10:13:03         CoffeePastryBread         1      0   

                                 kids  snacks  bread  breakfast_pastry  \
Transaction DateTime                                                     
1           2016-10-30 09:58:11     0       0      1                 0   
2           2016-10-30 10:05:34     0       0      2                 0   
3           2016-10-30 10:07:57     0       0      0                 0   
4           2016-10-30 10:08:41     0       0      0                 1   
5           2016-10-30 10:13:03     0       0      1                 1   

                                 dessert  condiments  breakfast  lunch  \
Transaction DateTime                                                     
1           2016-10-30 09:58:11        0           0          0      0   
2           2016-10-30 10:05:34        0           0          0      0   
3           2016-10-30 10:07:57        1           1          0      0   
4           2016-10-30 10:08:41        0           0          0      0   
5           2016-10-30 10:13:03        0           0          0      0   

                                 other_food  
Transaction DateTime                         
1           2016-10-30 09:58:11           0  
2           2016-10-30 10:05:34           0  
3           2016-10-30 10:07:57           0  
4           2016-10-30 10:08:41           0  
5           2016-10-30 10:13:03           0

Now the transaction number and the datetime are indices in this aggregated dataset. If we would like to use the information in these columns, we would have to reset the index.

In [29]:
bread_group.reset_index(level=['DateTime'], inplace=True)

Next, we will generate a column for day of week and for hour.

In [30]:
bread_group['hour'] = bread_group.DateTime.dt.hour
bread_group['day'] = bread_group.DateTime.dt.day_name()
bread_group.day.value_counts()

day
Saturday     2068
Friday       1488
Sunday       1264
Thursday     1252
Tuesday      1203
Monday       1135
Wednesday    1121
Name: count, dtype: int64

Saturday has the most transactions of any weekday by far.

In [31]:
bread_group.hour.value_counts()

hour
11    1445
12    1347
10    1267
13    1163
14    1130
9     1007
15     924
16     583
8      375
17     160
18      52
19      34
7       16
20      15
22       7
23       3
21       2
1        1
Name: count, dtype: int64

In [32]:
bread_days = pd.get_dummies(data=bread_group, columns=['day'])
bread_days.drop(columns='DateTime', inplace=True, axis=1)

In [33]:
bread_days

Date  \
Transaction                                             
1                                          2016-10-30   
2                                2016-10-302016-10-30   
3                      2016-10-302016-10-302016-10-30   
4                                          2016-10-30   
5                      2016-10-302016-10-302016-10-30   
...                                               ...   
9680                                       2017-04-09   
9681         2017-04-092017-04-092017-04-092017-04-09   
9682         2017-04-092017-04-092017-04-092017-04-09   
9683                             2017-04-092017-04-09   
9684                                       2017-04-09   

                                         Time  \
Transaction                                     
1                                    09:58:11   
2                            10:05:3410:05:34   
3                    10:07:5710:07:5710:07:57   
4                                    10:08:41   
5                    10:13:0310:13:0310:13:03   
...                                       ...   
9680                                 14:24:03   
9681         14:30:0914:30:0914:30:0914:30:09   
9682         14:32:5814:32:5814:32:5814:32:58   
9683                         14:57:0614:57:06   
9684                                 15:04:24   

                                                  Item  beverage  other  kids  \
Transaction                                                                     
1                                                Bread         0      0     0   
2                             ScandinavianScandinavian         0      0     0   
3                              Hot chocolateJamCookies         1      0     0   
4                                               Muffin         0      0     0   
5                                    CoffeePastryBread         1      0     0   
...                                                ...       ...    ...   ...   
9680                                             Bread         0      0     0   
9681         TrufflesTeaSpanish BrunchChristmas common         1      1     0   
9682                       MuffinTacos/FajitaCoffeeTea         2      0     0   
9683                                      CoffeePastry         1      0     0   
9684                                         Smoothies         1      0     0   

             snacks  bread  breakfast_pastry  dessert  ...  lunch  other_food  \
Transaction                                            ...                      
1                 0      1                 0        0  ...      0           0   
2                 0      2                 0        0  ...      0           0   
3                 0      0                 0        1  ...      0           0   
4                 0      0                 1        0  ...      0           0   
5                 0      1                 1        0  ...      0           0   
...             ...    ...               ...      ...  ...    ...         ...   
9680              0      1                 0        0  ...      0           0   
9681              0      0                 0        1  ...      0           1   
9682              0      0                 1        0  ...      0           1   
9683              0      0                 1        0  ...      0           0   
9684              0      0                 0        0  ...      0           0   

             hour  day_Friday  day_Monday  day_Saturday  day_Sunday  \
Transaction                                                           
1               9       False       False         False        True   
2              10       False       False         False        True   
3              10       False       False         False        True   
4              10       False       False         False        True   
5              10       False       False         False        True   
...           ...         ...         ...           ...         

## Some More Transformations - PCA

Our plan here is to use k-means clustering. However, an important note on k-means clustering is that it does not respond well to dummy variable columns. Therefore, our best option is to transform the data using principal component analysis or PCA. What PCA does is project our data onto a lower dimensional subspace. The new data will typically reduce the dimensions of our original data and will therefore, contain less variables. The first dimension will explain the most amount of variation in the data and subsequent components will explain less and less variation. This transformation will provide us with a smaller amount of continuous variables that we can cluster more effectively.

Here we chose to generate 4 components.

In [35]:
from sklearn.decomposition import PCA

pca = PCA(n_components=4)

# Select only numeric columns for PCA
numeric_columns = bread_days.select_dtypes(include=[np.number])
principalComponents = pca.fit_transform(numeric_columns)
principalDf = pd.DataFrame(data = principalComponents ,columns = ['pc1', 'pc2', 'pc3', 'pc4'])

principalDf.head()           

pc1       pc2       pc3       pc4
0 -3.197414 -0.949911  0.230320 -0.005280
1 -2.227886 -1.297669  1.179117 -0.106959
2 -2.107625  0.402060 -0.295612  0.836455
3 -2.198944 -0.603934 -0.811293 -0.114004
4 -2.206799  0.014565  0.435691 -0.332430

## The Algorithm - K-Means Clustering

We are now ready to cluster the data using k-means. We chose to create 5 clusters. The choice is normally arbitrary though there are ways to optimize the number of clusters. Here, the choice is more driven by the number of transaction clusters we would like to create. Two clusters would definitely be too few to capture meaningful differences while 10 is certainly too many.

In [36]:
from sklearn.cluster import KMeans

kmeans = KMeans(n_clusters=5)
bread_clusters = kmeans.fit(principalDf)
bread_clusters.cluster_centers_

c:\Users\happy\anaconda3_32\Lib\site-packages\joblib\externals\loky\backend\context.py:136: UserWarning: Could not find the number of physical cores for the following reason:
[WinError 2] The system cannot find the file specified
Returning the number of logical cores instead. You can silence this warning by setting LOKY_MAX_CPU_COUNT to the number of cores you want to use.
  warnings.warn(
  File "c:\Users\happy\anaconda3_32\Lib\site-packages\joblib\externals\loky\backend\context.py", line 257, in _count_physical_cores
    cpu_info = subprocess.run(
               ^^^^^^^^^^^^^^^
  File "c:\Users\happy\anaconda3_32\Lib\subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\happy\anaconda3_32\Lib\subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "c:\Users\happy\anaconda3_32\Lib\subprocess.py", line 1538, in _execute_child
    hp, ht, pid, tid = _w

array([[ 2.16838871e+00,  7.16314508e-01,  6.60485616e-02,
        -2.69635220e-02],
       [-2.84926563e+00, -8.93572434e-03, -1.70016398e-03,
         7.80059251e-04],
       [ 1.16333732e+00, -4.93193113e-01, -7.07641807e-02,
         2.22052050e-02],
       [ 4.00990169e+00, -3.83760371e-01, -1.03386859e-02,
         1.11925931e-02],
       [-6.78357584e-01,  1.10898855e-02,  4.49385653e-03,
        -1.36199070e-03]])

Let's apply the labels back to our original data so we can do some analysis.

In [38]:
breadbasket

Date      Time  Transaction           Item            DateTime  \
0      2016-10-30  09:58:11            1          Bread 2016-10-30 09:58:11   
1      2016-10-30  10:05:34            2   Scandinavian 2016-10-30 10:05:34   
2      2016-10-30  10:05:34            2   Scandinavian 2016-10-30 10:05:34   
3      2016-10-30  10:07:57            3  Hot chocolate 2016-10-30 10:07:57   
4      2016-10-30  10:07:57            3            Jam 2016-10-30 10:07:57   
...           ...       ...          ...            ...                 ...   
21288  2017-04-09  14:32:58         9682         Coffee 2017-04-09 14:32:58   
21289  2017-04-09  14:32:58         9682            Tea 2017-04-09 14:32:58   
21290  2017-04-09  14:57:06         9683         Coffee 2017-04-09 14:57:06   
21291  2017-04-09  14:57:06         9683         Pastry 2017-04-09 14:57:06   
21292  2017-04-09  15:04:24         9684      Smoothies 2017-04-09 15:04:24   

       beverage  other  kids  snacks  bread  breakfast_pastry  dessert  \
0             0      0     0       0      1                 0        0   
1             0      0     0       0      1                 0        0   
2             0      0     0       0      1                 0        0   
3             1      0     0       0      0                 0        0   
4             0      0     0       0      0                 0        0   
...         ...    ...   ...     ...    ...               ...      ...   
21288         1      0     0       0      0                 0        0   
21289         1      0     0       0      0                 0        0   
21290         1      0     0       0      0                 0        0   
21291         0      0     0       0      0                 1        0   
21292         1      0     0       0      0                 0        0   

       condiments  breakfast  lunch  other_food  
0               0          0      0           0  
1               0          0      0           0  
2               0          0      0           0  
3               0          0      0           0  
4               1          0      0           0  
...           ...        ...    ...         ...  
21288           0          0      0           0  
21289           0          0      0           0  
21290           0          0      0           0  
21291           0          0      0           0  
21292           0          0      0           0  

[21293 rows x 16 columns]

In [37]:
bread_days['labels'] = bread_clusters.fit_predict(principalDf)
bread_days.reset_index('Transaction', inplace=True)
bread_merged = pd.merge(breadbasket, bread_days[['Transaction', 'labels']], on='Transaction', how='outer')
bread_merged.head()

Date      Time  Transaction           Item            DateTime  \
0  2016-10-30  09:58:11            1          Bread 2016-10-30 09:58:11   
1  2016-10-30  10:05:34            2   Scandinavian 2016-10-30 10:05:34   
2  2016-10-30  10:05:34            2   Scandinavian 2016-10-30 10:05:34   
3  2016-10-30  10:07:57            3  Hot chocolate 2016-10-30 10:07:57   
4  2016-10-30  10:07:57            3            Jam 2016-10-30 10:07:57   

   beverage  other  kids  snacks  bread  breakfast_pastry  dessert  \
0         0      0     0       0      1                 0        0   
1         0      0     0       0      1                 0        0   
2         0      0     0       0      1                 0        0   
3         1      0     0       0      0                 0        0   
4         0      0     0       0      0                 0        0   

   condiments  breakfast  lunch  other_food  labels  
0           0          0      0           0       3  
1           0          0      0           0       3  
2           0          0      0           0       3  
3           0          0      0           0       3  
4           1          0      0           0       3

Let's so some analysis on the clusters. First let's look at how many transactions we have per cluster.

In [ ]:
bread_merged.labels.value_counts()

2    5296
3    4446
1    4041
0    3932
4    3578
Name: labels, dtype: int64

The largest cluster is the 5th cluster (our clusters are numbered 0 through 4).

One interesting thing to check is whether the clusters captured a different type of transaction by looking at the hour breakdown for each cluster.

In [39]:
pd.crosstab(bread_days.hour,bread_days.labels)

labels     0     1    2     3     4
hour                               
1          0     0    0     1     0
7          0     0    0    16     0
8          0     0    0   375     0
9          0     0    0  1007     0
10         0     0    0  1267     0
11      1445     0    0     0     0
12       478     0    0     0   869
13         0     0    0     0  1163
14         0  1130    0     0     0
15         0   924    0     0     0
16         0     0  583     0     0
17         0     0  160     0     0
18         0     0   52     0     0
19         0     0   34     0     0
20         0     0   15     0     0
21         0     0    2     0     0
22         0     0    7     0     0
23         0     0    3     0     0

We can clearly see a separation. Clusters 0, 2, and 4 center around noon. Cluster 1 is an early morning cluster. Cluster 3 is an evening cluster.

We can do the same analysis for day of week.

In [40]:
pd.crosstab(bread_group.day,bread_days.labels)

labels       0    1    2    3    4
day                               
Friday     330  332  127  378  301
Monday     193  238  115  358  231
Saturday   422  470  205  451  427
Sunday     252  260   88  333  299
Thursday   244  203  104  423  278
Tuesday    263  284   95  337  224
Wednesday  191  237  110  351  232

In cluster 1 (the early morning cluster), the disparity between the weekends and the weekdays is small. While in the clusters that center around later times, there seem to be more transactions during the weekends.

Let's also look at the top 5 products per cluster

In [41]:
a = bread_merged.groupby(['labels']).Item.value_counts()
b = a.to_frame("counts").reset_index()
b.set_index("Item", inplace=True)
b.groupby('labels').counts.nlargest(5)

labels  Item         
0       Coffee           1059
        Bread             957
        Tea               198
        Pastry            180
        Cake              157
1       Coffee           1155
        Bread             651
        Tea               440
        Cake              356
        Sandwich          246
2       Coffee            408
        Bread             250
        Tea               175
        Cake              154
        Hot chocolate      94
3       Coffee           1615
        Bread            1082
        Pastry            453
        Medialuna         294
        Tea               280
4       Coffee           1234
        Bread             385
        Sandwich          368
        Tea               342
        NONE              296
Name: counts, dtype: int64

While tea and coffee are popular in all 3 clusters, the morning cluster (Cluster 1) contains only bread, beverages and breakfast pastries. Clusters 0 and 2 are afternoon cluster and contain cake and sandwiches as top items. Cluster 3 is also an afternoon cluster and contains more desserts.

We can use this data to run promotions for certain items like cake and sandwiches at certain hours to increase our sales.